// @material/web/chips/lib/chip.js
/**
 * @license
 * Copyright 2023 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { __decorate } from "tslib";
import '../../elevation/elevation.js';
import '../../focus/focus-ring.js';
import '../../ripple/ripple.js';
import { html, LitElement, nothing } from 'lit';
import { property, queryAsync, state } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { html as staticHtml, literal } from 'lit/static-html.js';
import { pointerPress, shouldShowStrongFocus } from '../../focus/strong-focus.js';
import { ripple } from '../../ripple/directive.js';
/**
 * A chip component.
 */
export class Chip extends LitElement {
    constructor() {
        super(...arguments);
        this.disabled = false;
        this.elevated = false;
        this.href = '';
        this.label = '';
        this.target = '';
        this.showFocusRing = false;
        this.showRipple = false;
        this.getRipple = () => {
            this.showRipple = true;
            return this.ripple;
        };
    }
    render() {
        const button = this.href ? literal `a` : literal `button`;
        return staticHtml `
      <${button} class="container ${classMap(this.getContainerClasses())}"
          ?disabled=${this.disabled}
          href=${this.href || nothing}
          target=${this.href ? this.target : nothing}
          @blur=${this.handleBlur}
          @focus=${this.handleFocus}
          @pointerdown=${this.handlePointerDown}
          ${ripple(this.getRipple)}>
        ${!this.elevated ? html `<span class="outline"></span>` : nothing}
        ${this.elevated ? html `<md-elevation></md-elevation>` : nothing}
        ${this.showRipple ? this.renderRipple() : nothing}
        <md-focus-ring .visible=${this.showFocusRing}></md-focus-ring>
        <span class="icon leading">
          ${this.renderLeadingIcon()}
        </span>
        <span class="label">${this.label}</span>
        <span class="icon trailing">
          ${this.renderTrailingIcon()}
        </span>
      </${button}>
    `;
    }
    getContainerClasses() {
        return {
            disabled: this.disabled,
            elevated: this.elevated,
        };
    }
    renderLeadingIcon() {
        return html `<slot name="leading-icon"></slot>`;
    }
    renderTrailingIcon() {
        return nothing;
    }
    renderRipple() {
        return html `<md-ripple ?disabled=${this.disabled}></md-ripple>`;
    }
    handleBlur() {
        this.showFocusRing = false;
    }
    handleFocus() {
        this.showFocusRing = shouldShowStrongFocus();
    }
    handlePointerDown() {
        pointerPress();
        this.showFocusRing = shouldShowStrongFocus();
    }
}
__decorate([
    property({ type: Boolean })
], Chip.prototype, "disabled", void 0);
__decorate([
    property({ type: Boolean })
], Chip.prototype, "elevated", void 0);
__decorate([
    property()
], Chip.prototype, "href", void 0);
__decorate([
    property()
], Chip.prototype, "label", void 0);
__decorate([
    property()
], Chip.prototype, "target", void 0);
__decorate([
    state()
], Chip.prototype, "showFocusRing", void 0);
__decorate([
    state()
], Chip.prototype, "showRipple", void 0);
__decorate([
    queryAsync('md-ripple')
], Chip.prototype, "ripple", void 0);
